empty_cpu: Add functions that don't depend on Tensor

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #70618
* #70617
* #70616
* #70615
* #70614
* __->__ #70613
* #70612

This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

Differential Revision: [D33623682](https://our.internmc.facebook.com/intern/diff/D33623682)

peterbell10 committed:
empty_cpu: Add functions that don't depend on Tensor

This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33623682).

ngimel commented:
I'm getting internal error 
```
sparsenn_fused_operators.cpp:1286:29: error: no viable overloaded '='
[CONTEXT]           p_node->Output(1) = at::detail::empty_cpu(
```
what should the fix be?

peterbell10 committed:
Update on "empty_cpu: Add functions that don't depend on Tensor"


This refactors `at::detail::empty_cpu` to use only `TensorBase` so you
can construct tensors without including `Tensor.h`. It also adds a
`TensorOptions` version to reduce friction in operators moving from
the `at::empty` API.

Differential Revision: [D33623682](https://our.internmc.facebook.com/intern/diff/D33623682)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -386,7 +386,7 @@ struct TORCH_API IValue final {
   // we prefer to write them manually for clarity
 
-  IValue(at::Tensor t) : tag(Tag::Tensor), is_intrusive_ptr(false) {
+  IValue(at::TensorBase t) : tag(Tag::Tensor), is_intrusive_ptr(false) {
```
@ngimel this should be a more general solution. It allows `IValue` to be constructed from `TensorBase`.

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33623682).

ngimel commented:
I'm getting linker errors:
```
Summary: 
ld.lld: error: undefined symbol: at::detail::empty_cpu(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>)
stderr: ld.lld: error: undefined symbol: at::detail::empty_cpu(c10::ArrayRef<long>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>, c10::optional<c10::MemoryFormat>)
```

peterbell10 commented:
That functions is definitely defined so I'm guessing `EmptyTensor.cpp` isn't being linked for some reason. Maybe it needs be added to the fb-internal build definition?

ngimel commented:
looks like it needs to be added to tools/build_variables.bzl, let me try that
